### ISSUE TITLE
Support for https with dredd ?

### ISSUE THREAD

**Issue by erajkovic** [apiaryio/dredd#1010]
#### Describe your problem

Support for HTTPS / with proxy ?

From reading the doc, after finding the question to run with debug, looks as if the answer was no - would be nice to have this upfront as an english message in the tool.

"Dredd intentionally does not support HTTP(S) proxies for testing. Proxy can deliberately modify requests and responses or to behave in a very different way then the server under test. Testing over a proxy is, in the first place, testing of the proxy itself. That makes the test results irrelevant (and hard to debug)."

#### What command line options do you use?

```
$dredd https://apicatalog.oraclecloud.com/public/v1/orgs/oracle-public/apicollections/apicatalog/1.0/apis/Categories/canonical https://apicatalog.oraclecloud.com
```

#### What is in your `dredd.yml`?

```yaml
none
```

#### What's your `dredd --version` output?

```
error: 
Error when loading file from URL 'https://apicatalog.oraclecloud.com/public/v1/orgs/oracle-public/apicollections/apicatalog/1.0/apis/Categories/canonical'.
Is the provided URL correct?
```

#### Does `dredd --level=debug` uncover something?

verbose: 
HTTP(S) proxy specified by environment variables: NO_PROXY=localhost,127.0.0.1,erajkovi-mac,*.oracle.com,*.oraclecorp.com,/var/run/docker.sock, http_proxy=http://www-proxy.us.oracle.com:80, HTTPS_PROXY=http://www-proxy.us.oracle.com:80, https_proxy=http://www-proxy.us.oracle.com:80, no_proxy=localhost,127.0.0.1,erajkovi-mac,*.oracle.com,*.oraclecorp.com,/var/run/docker.sock, HTTP_PROXY=http://www-proxy.us.oracle.com:80. Please read documentation on how
Dredd works with proxies:
https://dredd.readthedocs.io/en/latest/how-it-works/#using-https-proxy

debug: Downloading https://apicatalog.oraclecloud.com/public/v1/orgs/oracle-public/apicollections/apicatalog/1.0/apis/Categories/canonical errored: Error: tunneling socket could not be established, cause=getaddrinfo ENOTFOUND www-proxy.us.oracle.com www-proxy.us.oracle.com:80

**Comment by honzajavorek** [comment]
You are right, Dredd does not support proxy for _testing_. It supports proxy for reporting, i.e. `--reporter=apiary`, and fetching the API description document from URL address, i.e. `dredd http://example.com/apiary.apib http://localhost:3000`.

> For corporate users, it may be good to have an option which enable proxy.

I understand corporate users often live behind a proxy, but the paragraph you quoted still holds.

When you test through proxy, you are testing the proxy instead of the API, and the test results become irrelevant. Imagine you have some headers in your specification, Dredd sends them, but the proxy strips them on the way. The request fails, although both Dredd and your API are fine. False negative, very hard to debug.

Worse example - imagine the proxy adds something, which accidentally makes your API compliant with the spec. Tests pass and you may think everything is fine. You deploy the API to production and make it public, with no proxy, and it won't work as intended. False positive. The whole purpose of Dredd gets beaten.

I'm fine leaving this issue open - let's discuss the topic, I can be missing something. But I'm yet to hear about a good way how to overcome the problem mentioned in the documentation.

**Comment by opichals** [comment]
@honzajavorek While acknowledging the dangers of through-proxy API testing I agree that this would indeed be a useful thing to have for people that have no other option but to use the proxy.

One could use [something like this](https://docs.mitmproxy.org/stable/concepts-howmitmproxyworks/#transparent-https) but that adds to the layers that may interfere.

**Comment by honzajavorek** [comment]
I am willing to think about a solution where the user is allowed to use proxy for testing, but Dredd's output shows a clear warning/disclaimer that using the proxy means testing the proxy itself, not the target API. Then it's up to the user whether they accept the risks or whether they know what they're doing.

**Comment by honzajavorek** [comment]
@erajkovic I'm glad you could get it working! And yes, beware running Dredd against live API instances 🙂

**Comment by honzajavorek** [comment]
> I am willing to think about a solution where the user is allowed to use proxy for testing, but Dredd's output shows a clear warning/disclaimer that using the proxy means testing the proxy itself, not the target API. Then it's up to the user whether they accept the risks or whether they know what they're doing.

I propose this as a solution to the problem. This isn't a priority issue though - I invite everyone interested in getting this implemented to add reaction to the original post so your desire to use this is [visible for us](https://github.com/apiaryio/dredd/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

<img width="316" alt="image" src="https://user-images.githubusercontent.com/283441/41159892-57b95052-6b2e-11e8-8c65-b69e4a1151cb.png">

**Comment by ertsaas** [comment]
I am running into this exact same issue. I have to run Dredd tests behind a proxy and would be willing to accept the risks that this entails, so in addition to my reaction on the original post, I am adding one more reaction here 👍 !
